README.md
# ZeroKit SDK for iOS

[![Build Status](https://api.travis-ci.org/tresorit/ZeroKit-iOS-SDK.svg?branch=master)](https://travis-ci.org/tresorit/ZeroKit-iOS-SDK)

[ZeroKit](https://tresorit.com/zerokit/) is a simple, breach-proof user authentication and end-to-end encryption library.

The ZeroKit SDK for iOS is currently under development and is accessible as a preview. We continuously improve the SDK and fix bugs.

You can [sign up for ZeroKit here](https://tresorit.com/zerokit/) and find [documentation here](https://tresorit.com/zerokit/docs/).

- [Requirements](#requirements)
- [Installation](#installation)
- [Usage](#usage)
- [Identity Provider](#identity-provider)
- [Administrative API](#administrative-api)
- [Example Application](#example-application)
- [Migration Guides](#migration-guides)
- [Known Issues and Limitations](#known-issues-and-limitations)
- [Third Party Libraries](#third-party-libraries)
- [Contact](#contact)
- [License](#license)

## Requirements
* Xcode 8.1+
* iOS 8.0+ (iOS 9.0+ to run the example app)
* Swift 3.0
* Objective-C compatible

## Installation
### Manually
After cloning the git repository or downloading the source code follow these steps:

1. Drag and drop the `ZeroKit.xcodeproj` into your Xcode project.
2. Add the `ZeroKit.xcodeproj/Products/ZeroKit.framework` to your application's Embedded Binaries.
3. Build your app.

### Carthage
[Carthage](https://github.com/Carthage/Carthage) is a decentralized dependency manager that builds your dependencies and provides you with binary frameworks.

To integrate ZeroKit into your Xcode project using Carthage, specify it in your Cartfile:

```
github "ZeroKit/ZeroKit"
```

Run `carthage update` to build the framework and drag the built ZeroKit.framework into your Xcode project.

### CocoaPods
[CocoaPods](https://cocoapods.org) is a dependency manager for Swift and Objective-C Cocoa projects.

To integrate ZeroKit into your Xcode project using CocoaPods, specify it in your Podfile:

```ruby
target '<Your Target>' do
  use_frameworks!
  pod 'ZeroKit', '~> 5.5'
end
```

Then, install it with the following command:

```
$ pod install
```

*Note: Pod installation may take a few minutes. The pod clones the ZeroKitClientNative crypto library which includes OpenSSL and builds these to link with the ZeroKit SDK.*

## Usage

### Initalizing ZeroKit
To initalize the SDK you will need your service URL. You can find your service URL in your [management portal](https://manage.tresorit.io).

*Swift*
```swift
import ZeroKit

let zeroKitConfig = ZeroKitConfig(serviceUrl: URL(string: "https://{TenantId}.api.tresorit.io")!)
let zeroKit = ZeroKit(config: zeroKitConfig)
```

*Objective-C*
```objc
#import <ZeroKit/ZeroKit-Swift.h>

NSURL *serviceURL = [NSURL URLWithString:@"https://{TenantId}.api.tresorit.io"];
ZeroKitConfig *config = [[ZeroKitConfig alloc] initWithServiceUrl:serviceURL];
ZeroKit *zeroKit = [[ZeroKit alloc] initWithConfig:config];
```

*Note: ZeroKit SDK for iOS wraps our javascript SDK. For this to run properly a web view is required to provide the runtime environment. As the user of the SDK you need not to interact with this web view and should not cause any trouble in most cases.*

### Password handling
A core concept of ZeroKit is that your application should not access and pass around the users' passwords. All password handling should be done by ZeroKit. For this we provide a `ZeroKitPasswordField` UI component you that you should present to users to enter their passwords. Some tips for using `ZeroKitPasswordField`:

- You can instantiate it in code or in Interface Builder. 
- It is marked as @IBDesignable so you can preview it in Interface Builder and also set some custom properties.
- In Interface Builder simply add a UIView to your layout and set its class to `ZeroKitPasswordField`. Also make sure the Module is set to `ZeroKit`.
- When you need the user to type their password twice for confirmation (eg. during registration) you create two password fields and set the `matchingField` property on them. Use the `passwordsMatch` property to check if they match.
- You can customize its appearance similarly to a `UITextField`.

To get the strength of a password use the `passwordStrength(passwordField:, completion:)` method of a `ZeroKit` object.

### Making a call

*Swift*
```swift
zeroKit.encrypt(plainText: "apple", inTresor: tresorId) { cipherText, error in
    guard error == nil else {
        print("Encrypting failed: \(error!)")
        return
    }

    print("Encrypted text: \(cipherText!)")
}
```

*Objective-C*
```objc
[self.zeroKit encryptWithPlainText:"apple" inTresor:tresorId completion:^(NSString * _Nullable cipherText, NSError * _Nullable error) {
    if (error) {
        NSLog(@"Encrypting failed: %@", error);
    } else {
        NSLog(@"Encrypted text: %@", cipherText);
    }
}];
```

## Identity Provider

ZeroKit comes with OpenID Connect provider implementation that you can use in your app. Use the `getIdentityTokens(clientId:completion:)` method of a `ZeroKit` object to get authorization code and identity token for the current ZeroKit user. A user must be logged in when calling this method.

You can add new SDK clients and edit settings on the [management portal](https://manage.tresorit.io).

## Administrative API

Most of the cryptographic operations (including invites and sharing) must be done client side by the SDK library. To provide control over these operations, and to prevent possible abuse by tampering the client, we introduced the admin API. All client initiated changes which has a permanent effect on the server has to be approved through the Admin API (typically by the server backend of the integrated app). For more information see the ZeroKit [documentation](https://tresorit.com/files/zerokit_encryption-sdk-documentation.pdf).

**IMPORTANT:** You must **never include your Admin key in your client application**. We provide a sample backend to handle the administrative calls for our sample app. The admin key must be kept secret and not included in any client applications that you distribute.

## Example Application

An example appliction is included with ZeroKit to demonstrate its usage. Open the `ZeroKit.xcworkspace` workspace which contains a `ZeroKitExample` project. It demonstrates the following features:

- Registration
- Login and logout
- Tresor creation
- Tresor sharing
- Encryption
- Decryption

### Configuring the Example

#### Step 1 - Set up the backend

The example app requires a backend to function. We created a sample backend that you can use for the mobile and web ZeroKit sample apps. You can find the backend and setup instructions [here](https://github.com/tresorit/ZeroKit-NodeJs-backend-sample).

#### Step 2 - Configure the iOS example app

To use the example app you first have to set it up with your ZeroKit configuration. Copy the sample `ZeroKitExample/Config.sample.plist` file and name it `Config.plist` in the same directory. In this `Config.plist` file set the values for the following keys:

- `ZeroKitServiceURL`: This is your tenant's service URL. You can find this URL on the [management portal](https://manage.tresorit.io).
- `ZeroKitClientId`: This is the client ID for your OpenID Connect client that you wish to use with your mobile. You can add new SDK clients and edit settings on the [management portal](https://manage.tresorit.io).
- `ZeroKitAppBackend`: This is the URL of the sample application backend that you set up in **Step 1**.

This is what the relevant section of the config file looks like:

```xml
<key>ZeroKitServiceURL</key>
<string>{ServiceUrl}</string>
<key>ZeroKitClientId</key>
<string>{ClientId}</string>
<key>ZeroKitAppBackend</key>
<string>{AppBackendUrl}</string>
```

#### Step 3 - Test Drive

Now you are ready to **Build and Run** (**⌘R**) the example app in Xcode.

### Registering Test Users

Register test users following the `'test-user-{XYZ}'` username format, eg. `test-user-Alice`. These users will be automatically validated by the sample backend so you can log in right after registration.

### Unit Tests

You can also take a look at the unit tests in this project to see further examples. To run the tests you must have the example app configured as described in [Configuring the Example](#configuring-the-example).

## Migration Guides

### Migration to 5.0.0 or later from 4.0.x or 4.1.x

#### Initialization

Since version 5.0.0 configuring ZeroKit with the Service URL is the preferred method. You can find your service URL on your [management portal](https://manage.tresorit.io).

```swift
// Since 5.0.0
let zeroKitConfig = ZeroKitConfig(serviceUrl: URL(string: "https://{TenantId}.api.tresorit.io")!)

// Previously in 4.1.x, deprecated
let zeroKitConfig = ZeroKitConfig(apiBaseUrl: URL(string: "https://{TenantId}.api.tresorit.io")!)

// Previously in 4.0.x, deprecated and removed
let zeroKitConfig = ZeroKitConfig(apiUrl: URL(string: "https://host-{HostId}.api.tresorit.io/tenant-{TenantId}/static/v4/api.html")!)
```

#### Notifications from 4.0.x

Since version 4.1.0 you no longer have to subscribe to `ZeroKit.DidLoadNotification` or `ZeroKit.DidFailLoadingNotification` notifications. You can start making ZeroKit calls right after initialization. The SDK will load the API internally and only after that will it handle your calls. This means you can just remove your observers for these notifications.

You will receive an error with `apiLoadingError` code in your callbacks if the API could not be loaded. In these cases the SDK will retry loading the API during subsequent calls.

Deprecated notifications are removed since 5.0.0.

## Known Issues and Limitations

Currently the SDK does not support app extensions.

## Third Party Libraries

* [ZeroKit Client Native Crypto](https://github.com/tresorit/ZeroKit-Client-Native-Crypto): ZeroKit client native cryptographic library.
* [LEB128](https://github.com/yageek/LEB128): Little Endian Base 128 with Swift.

## Contact

Contact us at [zerokit@tresorit.com](mailto:zerokit@tresorit.com).

## License

See the LICENSE file.


CHANGELOG.md
# Change Log

## [5.5.0](https://github.com/tresorit/ZeroKit-iOS-SDK/releases/tag/5.5.0)
### Changes
- ZeroKit has been rewritten to use the JavaScriptCore framework to execute javascript instead of WKWebView.
- Javascript execution is now possible while the app is in the background.
- Cryptographic operations are implemented in native code providing improved performance over javascript implementations.

### Added
- New error codes.
- More descriptive error messages for developers. This includes descriptions and origin of errors.
- You can now set the log level with `ZeroKit.logLevel`. Defaults to `warning`.

## [5.0.0](https://github.com/tresorit/ZeroKit-iOS-SDK/releases/tag/5.0.0) - 2017-07-18
### Changes
- Use ZeroKit JavaScript API version 5.
- `init(apiBaseUrl: URL)` on `ZeroKitConfig` has been deprecated in favor of `init(serviceUrl: URL)`. The value for the URL parameter is the same, it has been renamed for consistency.

### Added
- Invitation links can now be revoked using the `ZeroKit.revokeInvitationLink(forTresor:secret:completion:)` instance method.

### Removed
- Removed deprecated APIs: `ZeroKit.DidFailLoadingNotification`, `ZeroKit.DidLoadNotification`, `ZeroKit.isLoaded`, `ZeroKitConfig.init(apiUrl:)`.

## [4.1.2](https://github.com/tresorit/ZeroKit-iOS-SDK/releases/tag/4.1.2) - 2017-06-26
### Added
- `clear` method to `ZeroKitPasswordField`.

### Changes
- Improved data escaping performance during encryption and decryption.

### Example App Changes
- Added support for user public profile to the `Backend` class.
- Turned off autocorrection in username text fields.

## [4.1.1](https://github.com/tresorit/ZeroKit-iOS-SDK/releases/tag/4.1.1) - 2017-04-05
### Fixed
- Identity provider redirect URL when configured ZeroKit with a service URL ending with a slash or with deprecated `apiUrl` method.

### Example App Changes
- The example app now uses our [sample backend](https://github.com/tresorit/ZeroKit-NodeJs-backend-sample). See the updated [README](README.md) for more information.

## [4.1.0](https://github.com/tresorit/ZeroKit-iOS-SDK/releases/tag/4.1.0) - 2017-03-30
### Added
- OpenID Connect identity provider support. See the `getIdentityTokens` function on `ZeroKit`.
- ZeroKit can now be configured with the api base URL (also known as the service URL), `ZeroKitConfig(apiBaseUrl: URL(string: "https://{TenantID}.api.tresorit.io")!)`

### Changes
- In version 4.1.0 you no longer have to subscribe to `ZeroKit.DidLoadNotification` or `ZeroKit.DidFailLoadingNotification` notifications. You can start making ZeroKit calls right after initialization. The SDK will load the API internally and only after that will it handle your calls. This means you can just remove your observers for these notifications.
- *Deprecated:* Due to the above changes, `ZeroKit.DidLoadNotification`, `ZeroKit.DidFailLoadingNotification` and `ZeroKit.isLoaded` have been deprecated and will be removed in a future version.
- *Deprecated:* `ZeroKitConfig(apiUrl: URL)` has been deprecated in favor of `ZeroKitConfig(apiBaseUrl: URL)` and will be removed in a future version.
- See the migration guide section in the [README](README.md) for more information on the changes.

## [4.0.3](https://github.com/tresorit/ZeroKit-iOS-SDK/releases/tag/4.0.3) - 2017-03-24
### Added
- Password strength estimation to `ZeroKit`. You can pass a `ZeroKitPasswordField` directly to `ZeroKit.passwordStrength(passwordField:completion:)`.

### Changes
- *Deprecated:* `ZeroKitPasswordField.isPasswordValid` has been deprecated and will be removed in a future version. Use the above mentioned password strength method instead.

## [4.0.2](https://github.com/tresorit/ZeroKit-iOS-SDK/releases/tag/4.0.2) - 2017-02-13
### Added
- Tests for two simultaneous ZeroKit users.

### Example App Changes
- Changed the api URL format in the example app configuration.

## [4.0.1](https://github.com/tresorit/ZeroKit-iOS-SDK/releases/tag/4.0.1) - 2017-01-20
First public release.


ZeroKit/ZeroKitError.swift
import Foundation
import JavaScriptCore


/**
 Errors returned by ZeroKit.
 */
@objc public enum ZeroKitError: Int, Error {
    
    /** An unknown error occurred. */
    case unknownError = 1
    
    /** The result received is of unexpected type or format. */
    case unexpectedResult = 2
    
    /** The data was not in the expected format. */
    case dataFormatError = 3
    
    /** ZeroKit iOS SDK uses web view for the identity provider (IDP) funcionality. It performs the operation in a web view that is added as the bottom most view to the app's window. In most cases it should not cause any trouble. */
    case cannotAddWebView = 4
    
    /** The provided password is empty. */
    case passwordIsEmpty = 5
    
    /** Login by 'remember be' is not possible. Eg. the user did not log in previously with 'remember me'. */
    case cannotLoginByRememberMe = 6
    
    /** Invalid authorization, ie. the password is incorrect. */
    case invalidAuthorization = 7
    
    /** User with the specified user ID does not exist. */
    case userDoesNotExist = 8
    
    /** The user is already member of the tresor. */
    case alreadyMember = 9
    
    /** The user is not member of the tresor. */
    case notMember = 10
    
    /** The caller user is not member of the tresor. */
    case callerUserIsNotMemberOfTresor = 11
    
    /** Received invalid input. */
    case badInput = 12
    
    /** Invalid user ID was provided. */
    case invalidUserId = 13
    
    /** A network error occurred. */
    case networkError = 14
    
    /** The response is not valid, ie. response verification failed. */
    case invalidResponse = 15
    
    /** The request is not valid. */
    case invalidRequest = 16
    
    /** An internal error occurred. */
    case internalError = 17
    
    /** User is not logged in. Login is required to perform the operation. */
    case loginRequired = 18
    
    /** Failed to load ZeroKit API. This can happen, for example, when a network error occurs. You can retry the request when you get this error. */
    case apiLoadingError = 19
    
    /** Request was interrupted by the user. */
    case userInterrupted = 20
    
    /** Could not find tresor by ID. */
    case tresorNotExists = 21
    
    /** This tresor has been deleted or rejected at creation. */
    case tresorAlreadyDeleted = 22
    
    /** This tresor is not yet approved. */
    case tresorIsNotApproved = 23
    
    /** The user is not logged in. */
    case notLoggedInError = 24
    
    /** There is no user by that ID. */
    case userNotFound = 25
    
    /** You cannot invite yourself to a tresor. */
    case cantInviteYourself = 26
    
    /** You cannot kick yourself from a tresor. */
    case cantKickYourself = 27
    
    /** Item not found for operation */
    case notFound = 28
    
    /** Operation was not allowed. */
    case forbidden = 29
    
    /** The user is not validated. */
    case userNotValidated = 30
    
    /** There is a problem with the provided registration session ID. */
    case regSessionNotExists = 31
    
    /** The registration session and user ID mismatch. */
    case userIdMismatch = 32
    
    /** Bad password try limit exceeded. */
    case badPasswordTryLimitExceeded = 33
    
    /** User is already registered. */
    case userAlreadyRegistered = 34
    
    /** User does not exist */
    case userNotExists = 35
    
    fileprivate static func errorCodeFromJavascript(_ value: JSValue) -> (ZeroKitError, String?) {
        if let dict = value.toDictionary() {
            return self.errorCodeFromDictionary(dict)
        }
        return (.unknownError, nil)
    }
    
    fileprivate static func errorCodeFromDictionary(_ value: [AnyHashable: Any]) -> (ZeroKitError, String?) {
        if let code = value["code"] as? String {
            let description = self.description(from: value as? [NSObject: NSObject])
            
            if code == "InternalError" {
                if let internalException = value["internalException"] as? [AnyHashable: Any],
                    let internalCode = internalException["code"] as? String {
                    return (error(from: internalCode), description)
                }
            }
            return (error(from: code), description)
        }
        return (.unknownError, nil)
    }
    
    private static func description(from dict: [NSObject: NSObject]?) -> String? {
        guard let dict = dict else {
            return nil
        }
        
        var entries = [String]()
        for (key, value) in dict {
            if let value = value as? [NSObject: NSObject] {
                entries.append(String(format: "%@: %@", key, self.description(from: value)!))
            } else {
                entries.append(String(format: "%@: %@", key, value))
            }
        }
        return String(format: "{ %@ }", entries.joined(separator: ", "))
    }
    
    private static let errorMap: [String: ZeroKitError] = [
        "AlreadyAMember": .alreadyMember,
        "AlreadyMember": .alreadyMember,
        "BadInput": .badInput,
        "BadPasswordTryLimitExceeded": .badPasswordTryLimitExceeded,
        "CallerUserIsNotMemberOfTresor": .callerUserIsNotMemberOfTresor,
        "CantInviteYourself": .cantInviteYourself,
        "CantKickYourself": .cantKickYourself,
        "Forbidden": .forbidden,
        "InternalError": .internalError,
        "InvalidAuthorization": .invalidAuthorization,
        "InvalidUserId": .invalidUserId,
        "NotFound": .notFound,
        "NotLoggedInError": .notLoggedInError,
        "NotMember": .notMember,
        "RegSessionNotExists": .regSessionNotExists,
        "TresorAlreadyDeleted": .tresorAlreadyDeleted,
        "TresorIsNotApproved": .tresorIsNotApproved,
        "TresorNotExists": .tresorNotExists,
        "UserAlreadyRegistered": .userAlreadyRegistered,
        "UserIdMismatch": .userIdMismatch,
        "UserNameDoesntExist": .userDoesNotExist,
        "UserNotExists": .userNotExists,
        "UserNotFound": .userNotFound,
        "UserNotValidated": .userNotValidated,
    ]
    
    fileprivate static func error(from stringCode: String) -> ZeroKitError {
        if let error = errorMap[stringCode] {
            return error
        }
        Log.v("Unexpected error code: %@", stringCode)
        return .unknownError
    }
    
    public static func ==(lhs: ZeroKitError, rhs: Error) -> Bool {
        guard let zkRhs = rhs as? ZeroKitError else {
            return false
        }
        
        return lhs.rawValue == zkRhs.rawValue
    }
    
    // MARK: Format test support
    static func errorCodeString(for error: NSError) -> String {
        var inverseMap = [ZeroKitError: String]()
        for (key, value) in ZeroKitError.errorMap {
            inverseMap[value] = key
        }
        if let zkError = error as? ZeroKitError,
            let code = inverseMap[zkError] {
            return code
        }
        return "UnexpectedError \(error)"
    }
}


public func ==<T: Error>(lhs: T, rhs: ZeroKitError) -> Bool {
    return rhs == lhs
}


extension NSError {
    convenience init(_ result: Any?, defaultErrorCode: ZeroKitError = .unknownError, message: String? = nil, line: Int = #line, file: String = #file) {
        var code = defaultErrorCode
        let domain = (ZeroKitError.unknownError as NSError).domain
        var info = [AnyHashable: Any]()
        var description: String?
        
        func handle(error: NSError) {
            if let zkError = error as? ZeroKitError {
                code = zkError
            }
            info[NSUnderlyingErrorKey] = error as NSError
        }
        
        if let error = result as? NSError {
            handle(error: error)
        } else if let value = result as? JSValue {
            if let error = value.toObject() as? NSError {
                handle(error: error)
            } else {
                (code, description) = ZeroKitError.errorCodeFromJavascript(value)
            }
        } else if let value = result as? [AnyHashable: Any] {
            (code, description) = ZeroKitError.errorCodeFromDictionary(value)
        } else if let value = result as? String {
            code = ZeroKitError.error(from: value)
            description = value
        }
        
        if let message = message {
            info["ZeroKitErrorMessage"] = message
        }
        if let description = description {
            info["ZeroKitErrorDescription"] = description
        }
        info["ZeroKitErrorOrigin"] = String(format: "%@:%d", (file as NSString).lastPathComponent, line)
        
        self.init(domain: domain, code: code.rawValue, userInfo: info)
        
        Log.v("Error created: %@", self.description)
    }
}
